dynet/dim.cc

```diff
@@ -7,7 +7,7 @@ using namespace std;
 namespace dynet {
 
 ostream& operator<<(ostream& os, const Dim& d) {
-  os << '{';
+  os << d.nd << '-' << '{';
```

Comment:
I think this may be confusing when people try to print out `Dim` for debugging purposes, etc. Could we leave `<<` the same way as before and modify `>>` to automatically infer `nd` as necessary?